proposals/records.md

```diff
@@ -39,49 +39,119 @@ The synthesized members are as follows:
 ```C#
-protected virtual Type EqualityContract { get; }
+public virtual bool Equals(R? other);
 ```
-
-It is an error if the base implementations of any of the overridden members is sealed or non-virtual,
-or do not match the expected signature and accessibility.
-
-`Equals(T other)` returns true if and only if each of the following terms are true:
+The synthesized `Equals(R?)` returns `true` if and only if each of the following are `true`:
 - `other` is not `null`, and
-- For each field declared in the record type, the value of
+- For each instance field in the record type that is not inherited, the value of
 `System.Collections.Generic.EqualityComparer<TN>.Default.Equals(fieldN, other.fieldN)` where `TN` is the field type, and
-- If there is a base record type, the value of `base.Equals(other)`; otherwise
+- If there is a base record type, the value of `base.Equals(other)` (a non-virtual call to `public virtual bool Equals(Base? other)`); otherwise
 the value of `EqualityContract.Equals(other.EqualityContract)`.
 
-The overrides of `Equals(T other)` for the base methods, including `object.Equals(object other)`, perform the equivalent of:
+If the record type is derived from a base record type `Base`, the record type includes a synthesized override of the strongly-typed `Equals(Base other)`.
+The synthesized override is `sealed`.
+It is an error if the override is declared explicitly.
+The synthesized override returns `Equals((object?)other)`.
+
+The record type includes a synthesized override of `object.Equals(object? obj)`.
+It is an error if the override is declared explicitly.
+The synthesized override returns `Equals(other as R)` where `R` is the record type.
+```C#
+public override bool Equals(object? obj);
+```
+
+The record type includes a synthesized override of `object.GetHashCode()`.
+The method can be declared explicitly.
+It is an error if the explicit declaration is `sealed` unless the record type is `sealed`.
 ```C#
-public override bool Equals(object other) => Equals(other as T);
+public override int GetHashCode();
 ```
+A warning is reported if one of `Equals(R?)` and `GetHashCode()` is explicitly declared but the other method is not explicit.
 
-`GetHashCode()` returns the `int` result of a deterministic function taking the following values:
-- For each field declared in the record type, the value of
+The synthesized override of `GetHashCode()` returns an `int` result of a deterministic function combining the following values:
+- For each instance field in the record type that is not inherited, the value of
 `System.Collections.Generic.EqualityComparer<TN>.Default.GetHashCode(fieldN)` where `TN` is the field type, and
 - If there is a base record type, the value of `base.GetHashCode()`; otherwise
 the value of `System.Collections.Generic.EqualityComparer<System.Type>.Default.GetHashCode(EqualityContract)`.
```

Comment:
>value [](start = 4, length = 5)

Since you don't guard against a null `EqualityContract` in `Equals`, you needn't do so here. You should do so in both or neither.

Reply:
Changed `Equals` to use `EqualityContract == other.EqualityContract`.

---
In reply to: [442963587](https://github.com/dotnet/csharplang/pull/3583#discussion_r442963587) [](ancestors = 442963587)